common/djangoapps/student/models.py

```diff
@@ -199,6 +200,61 @@ def user_by_anonymous_id(uid):
         return None
 
 
+def get_retired_username_by_username(username):
+    """
+    Returns a "retired username" hashed using the newest configured salt
+    """
+    return user_util.get_retired_username(username, settings.RETIRED_USER_SALTS, settings.RETIRED_USERNAME_FMT)
+
+
+def get_retired_email_by_email(email):
+    """
+    Returns a "retired email" hashed using the newest configured salt
+    """
+    return user_util.get_retired_email(email, settings.RETIRED_USER_SALTS, settings.RETIRED_EMAIL_FMT)
+
+
+def get_all_retired_usernames_by_username(username):
+    """
+    Returns a generator of "retired usernames", one hashed with each
+    configured salt. Used for finding out if the given username has
+    ever been used and retired.
+    """
+    return user_util.get_all_retired_usernames(username, settings.RETIRED_USER_SALTS, settings.RETIRED_USERNAME_FMT)
+
+
+def get_all_retired_emails_by_email(email):
+    """
+    Returns a generator of "retired emails", one hashed with each
+    configured salt. Used for finding out if the given email has
+    ever been used and retired.
+    """
+    return user_util.get_all_retired_emails(email, settings.RETIRED_USER_SALTS, settings.RETIRED_EMAIL_FMT)
+
+
+def get_potentially_retired_user_by_username_and_hash(username, hashed_username):
+    """
+    To assist in the retirement process this method will:
+    - Confirm that any locally hashed username matches the passed in one
+      (in case of salt mismatches with the upstream script).
+    - Attempt to return a User object based on the username, or if it
+      does not exist, the any hashed username salted with the historical
+      salts.
+    """
+    locally_hashed_usernames = list(get_all_retired_usernames_by_username(username))
+
+    if hashed_username not in locally_hashed_usernames:
+        raise Exception('Mismatched hashed_username, bad salt?')
+
+    try:
+        return User.objects.get(username=username)
+    except User.DoesNotExist:
+        # The 2nd DoesNotExist will bubble up from here if necessary,
+        # an assumption is being made here that our hashed username format
+        # is something that a user cannot create for themselves.
+        return User.objects.get(username__in=locally_hashed_usernames)
+
+
```

Comment:
What about the call that learner registration will use to check to see if a proposed username is already retired? Is it yet to be written?

Reply:
Is there a reason they can't use this one?

Reply:
Hmm - guess they can if it's called like this:
```
get_potentially_retired_user_by_username_and_hash(proposed_username, get_retired_username_by_username(proposed_username))
```
Is that correct? If so, I guess I was talking about a version where you simply passed in `proposed_username` to see if a `is_retired_username()` method returns True/False.

Reply:
Ah, I thought this was a comment on `get_all_retired_usernames_by_username` which could then be used to do the same filter as above, but I can add a convenience method for it. 👍